src/CodeGen_RISCV.cpp

```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+        }
+        Type arg_type = concretize_fixed_or_scalable(arg_type_pattern, bit_width_scale, target.vector_bits);
+
+        bool scalable = (arg_type_pattern.type_pattern != FixedOrScalableVector::Fixed);
+        mangled_name += "." + mangle_vector_argument_type(arg_type, scalable, effective_vscale);
+        llvm::Type *llvm_type;
+        if (arg_type.is_vector()) {
+            int lanes = arg_type.lanes();
+            if (scalable) {
+                lanes /= effective_vscale;
+            }
+            llvm_type = llvm::VectorType::get(llvm_type_of(arg_type.element_of()),
+                                              lanes, scalable);
+        } else {
+            llvm_type = llvm_type_of(arg_type);
+        }
+        llvm_arg_types.push_back(llvm_type);
+    }
+    if (intrin.flags & RISCVIntrinsic::AddVLArg) {
+        mangled_name += (target.bits == 64) ? ".i64" : ".i32";
+        llvm_arg_types.push_back(xlen_type);
+    }
+
+    llvm::Function *inner =
+        get_llvm_intrin(llvm_ret_type, mangled_name, llvm_arg_types);
+    llvm::FunctionType *inner_ty = inner->getFunctionType();
+
+    // Remove vector tail preservation argument.
+    llvm_arg_types.erase(llvm_arg_types.begin());
+    // Remove vector length argument passed to inrinsic for wrapper.
+    // Wrapper will supply a constant for the fixed vector length.
+    if (intrin.flags & RISCVIntrinsic::AddVLArg) {
+        llvm_arg_types.resize(llvm_arg_types.size() - 1);
+    }
+
+    string wrapper_name = unique_name(std::string(intrin.name) + "_wrapper");
+    llvm::FunctionType *wrapper_ty = llvm::FunctionType::get(
+        inner_ty->getReturnType(), llvm_arg_types, false);
+    llvm::Function *wrapper =
+        llvm::Function::Create(wrapper_ty, llvm::GlobalValue::InternalLinkage,
+                               wrapper_name, module.get());
+    llvm::BasicBlock *block =
+        llvm::BasicBlock::Create(module->getContext(), "entry", wrapper);
+    llvm::IRBuilderBase::InsertPoint here = builder->saveIP();
+    builder->SetInsertPoint(block);
+
+    // Set vector fixed-point rounding flag if needed for intrinsic.
+    bool round_down = intrin.flags & RISCVIntrinsic::RoundDown;
+    bool round_up = intrin.flags & RISCVIntrinsic::RoundUp;
+    if (round_down || round_up) {
+        internal_assert(!(round_down && round_up));
+        llvm::Value *rounding_mode = llvm::ConstantInt::get(xlen_type, round_down ? 2 : 0);
+        // TODO: When LLVM finally fixes the instructions to take rounding modes,
+        // this will have to change to passing the rounding mode to the intrinsic.
+        llvm::FunctionType *csrw_llvm_type = llvm::FunctionType::get(void_t, {xlen_type}, false);
+        llvm::InlineAsm *inline_csrw = llvm::InlineAsm::get(csrw_llvm_type, "csrw vxrm,${0:z}", "rJ,~{memory}", true);
+        builder->CreateCall(inline_csrw, {rounding_mode});
+    }
+
+    // Call the LLVM intrinsic.
+    // TODO: This does not work for reductions.
```

Comment:
Should we open an Issue for this? Or add `internal_assert(not-a-reduction);`?

Reply:
Reductions are now handled through the vector predication intrinsics, hence the reason I had that TODO there is obsolete. (It was tracking that I needed to figure out a way to make it work.) There are likely more to be handled, but it doesn't really need an assert as there's no way to specify an intrinsic that does a reduction in this part of the code. This is a thing to consider when revamping ```call_intrin```, but I don't think it needs a tracking issue as there is an ```is_reduction``` argument now. (I.e. I probably fixed the underlying issue there too.)